spotify.go

```diff
@@ -107,14 +124,48 @@ func (e Error) Error() string {
 }
 
 // decodeError decodes an Error from an io.Reader.
-func decodeError(r io.Reader) error {
+func decodeError(c *Client, resp *http.Response) error {
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	buf := bytes.NewBuffer(responseBody)
+
 	var e struct {
 		E Error `json:"error"`
 	}
-	err := json.NewDecoder(r).Decode(&e)
+	err = json.NewDecoder(buf).Decode(&e)
 	if err != nil {
-		return errors.New("spotify: couldn't decode error")
+		if len(responseBody) == 0 {
+			errors.New("spotify: could not decode empty body")
```

Comment:
- You're not returning an error here, simply creating one and throwing it away.
- It seems like you could check for an empty response body before trying to decode JSON.
- Since you're already checking for an empty body ahead of time, there's no need to put `len(responseBody)` in the error message
- Let's try to return the HTTP status code in the error if we can't decode it